rllib/offline/estimators/tests/test_ope.py

```diff
@@ -199,22 +217,440 @@ def test_fqe_model(self):
-        tmp_probs = torch.stack(tmp_probs).transpose(0, 1)
-        tmp_probs = convert_to_numpy(tmp_probs)
+            tmp_probs.append(np.exp(log_probs))
+        tmp_probs = np.stack(tmp_probs).T
         check(action_probs, tmp_probs, decimals=3)
 
-    def test_multiple_inputs(self):
-        # TODO (Rohan138): Test with multiple input files
-        pass
+    def test_fqe_optimal_convergence(self):
+        # Optimal CliffWalkingWallPolicy with epsilon = 0.0
+        # and CliffWalkingWallEnv are deterministic;
+        # check that FQE converges to the true Q-values for self.batch
+        q_vals = [
+            -2.50,
+            -1.51,
+            -0.52,
+            0.49,
+            1.50,
+            2.53,
+            3.56,
+            4.61,
+            5.67,
+            6.73,
+            7.81,
+            8.90,
+            10.00,
+        ]
+        q_model_config = {
+            "tau": 1.0,
+            "model": {
+                "fcnet_hiddens": [],
+                "activation": "linear",
+            },
+            "lr": 0.01,
+            "n_iters": 5000,
+            "delta": 1e-3,
+        }
+
+        fqe = FQETorchModel(
+            policy=self.policy,
+            gamma=self.gamma,
+            **q_model_config,
+        )
+        losses = fqe.train(self.batch)
+        print(losses[-10:])
+        assert losses[-1] < fqe.delta, "FQE loss did not converge!"
+        estimates = fqe.estimate_v(self.batch)
+        print(estimates)
+        check(
+            estimates,
+            q_vals,
+            decimals=1,
+        )
+
+
+def get_cliff_walking_wall_policy_and_data(
+    num_episodes: int,
+    gamma: float,
+    epsilon: float,
+) -> Tuple[Policy, SampleBatch, float, float]:
+    """Collect a cliff_walking_wall policy and data with epsilon-greedy exploration.
+
+    Args:
+        num_episodes: Number of episodes to collect
+        gamma: discount factor
+        epsilon: epsilon-greedy exploration value
+
+    Returns:
+        A Tuple consisting of:
+          - A CliffWalkingWallPolicy with exploration parameter epsilon
+          - A SampleBatch of `num_episodes` CliffWalkingWall episodes collected using
+          epsilon-greedy exploration
+          - The mean of the discounted return over the collected episodes
+          - The stddev of the discounted return over the collected episodes
+
+    """
+    config = (
+        AlgorithmConfig()
+        .rollouts(batch_mode="complete_episodes")
+        .environment(disable_env_checking=True)
+        .experimental(_disable_preprocessor_api=True)
+    )
+
+    env = CliffWalkingWallEnv()
+    policy = CliffWalkingWallPolicy(
+        env.observation_space, env.action_space, {"epsilon": epsilon}
+    )
+    workers = WorkerSet(
+        env_creator=lambda env_config: CliffWalkingWallEnv(),
+        policy_class=CliffWalkingWallPolicy,
+        trainer_config=config.to_dict(),
+        num_workers=8,
```

Comment:
could you not specify num_workers? The machine that we run these tests on may not have 8+ cpus if you don't specify fractional resources.

Reply:
**TODO**

Reply:
I don't know if there's a good way to avoid specifying num_workers-we'd still like to do parallelized sampling, but we're using the raw WorkerSet so we don't know how many workers we actually have. I spoke to Avnish yesterday and set it to 4 to be safe.